File: source/downloads/code/LeetCode/Python/Count-Numbers-with-Unique-Digits.py
class Solution(object):
    def countNumbersWithUniqueDigits(self, n):
        """
        :type n: int
        :rtype: int
        """

        # 长度位i的数字中不同digit的数量
        nums = [9] 
        for x in range(9,0,-1):
            nums.append(nums[-1] * x)
        # nums[3]不包括nums[2]
        return sum(nums[:n])+1

File: downloads/code/LeetCode/Java/Count-Numbers-with-Unique-Digits.java
public class Solution {
    public int countNumbersWithUniqueDigits(int n) {
        int[] dp = new int[11];
        if(n==0) return 1;
        if (n >10)
            return 0;
        dp[1] = 10;
        dp[2] = 9*9;
        for(int i = 3; i<=10; i++){
            dp[i] = dp[i-1]*(11-i);
        }
        // Get f[1]+f[2]+...f[n]
        int ret = 0;
        for (int i=0;i<=n;i++)
            ret += dp[i];
        return ret;
    }
}
